lib/storage/metadata/mongoclient/MongoClientInterface.js

```diff
@@ -1295,25 +1316,112 @@ class MongoClientInterface {
-                    { error: err.message });
+                    { error: err.message, bucket: bucketName, key: objName });
                 return cb(errors.InternalError);
             }
-            if (result.ok !== 1) {
+            if (res.ok !== 1) {
                 log.error(
-                    'deleteObjectNoVer: failed deleting object with no version',
-                    { error: err.message });
+                    'deleteObjectNoVer: failed to delete object',
+                    { bucket: bucketName, object: objName });
                 return cb(errors.InternalError);
             }
             return cb(null);
         });
     }
 
+    /**
+     * Flags the object before deleting it, this is done
+     * to keep object metadata in the oplog, as oplog delete
+     * events don't contain any object metadata
+     * @param {Object} collection MongoDB collection
+     * @param {string} bucketName bucket name
+     * @param {string} key Key of the object to delete
+     * @param {object} filter additional query filters
+     * @param {Logger}log logger instance
+     * @param {Function} cb callback containing error
+     * and BulkWriteResult
+     * @return {undefined}
+     */
+    internalDeleteObject(collection, bucketName, key, filter, log, cb) {
+        // filter used when finding and updating object
+        const findFilter = Object.assign({
+            _id: key,
+            $or: [
+                { 'value.deleted': { $exists: false } },
+                { 'value.deleted': { $eq: false } },
+            ],
+        }, filter);
+        // filter used when deleting object
+        const updateDeleteFilter = Object.assign({
+            '_id': key,
+            'value.deleted': true,
+        }, filter);
+        async.waterfall([
+            // Adding delete flag when getting the object
+            // to avoid having race conditions.
+            next => collection.findOneAndUpdate(findFilter, {
+                $set: {
+                    '_id': key,
+                    'value.deleted': true,
+                },
+            }, {
+                upsert: false,
+                returnDocument: 'after',
+            }, (err, doc) => {
+                if (err) {
+                    log.error('internalDeleteObject: error getting object',
+                        { bucket: bucketName, object: key, error: err.message });
+                    return next(errors.InternalError);
+                }
+                if (!doc.value) {
+                    log.error('internalDeleteObject: unable to find target object to delete',
+                        { bucket: bucketName, object: key });
+                    return next(errors.NoSuchKey);
+                }
+                const obj = doc.value;
+                const objMetadata = new ObjectMD(obj.value);
+                objMetadata.setOriginOp('s3:ObjectRemoved:Delete');
+                objMetadata.setDeleted(true);
```

Comment:
should already be set?
or we need to add it, since it is not "known" by ObjectMD (and it should be added there, possibly as an "optional" field,  to make sure the model is actually representative)

Reply:
The mongo memory server we use for our functional tests don't support the `returnDocument: 'after'` argument for some reason even when explicitly setting the mongo version to 4.0.27. so it always returns the version of the object before the update

Reply:
it is weird that it does not work... we may accept it, but then the code still seamingly depends on `returnDocumet:'after'` (requested line 1377) : so either we need to fix this, or do not even try to request the updated document (or at least at a comment and create a TechDebt ticket)